homeassistant/components/surepetcare/binary_sensor.py

```diff
@@ -0,0 +1,212 @@
+                               discovery_info=None):
+    """Set up Sure PetCare Flaps sensors based on a config entry."""
+    from surepy import SurePetcare
+
+    entities = list()
+    surepy: SurePetcare = hass.data[DATA_SURE_PETCARE][DATA_SUREPY]
+
+    for thing in hass.data[DATA_SURE_PETCARE][SURE_IDS]:
+        sure_id = thing[CONF_ID]
+        sure_type = thing[CONF_TYPE]
+
+        if sure_id not in hass.data[DATA_SURE_PETCARE][sure_type]:
+
+            if sure_type == SureThingID.FLAP.name:
+                hass.data[DATA_SURE_PETCARE][
+                    sure_type][sure_id] = await surepy.get_flap_data(sure_id)
+                entities.append(Flap(sure_id, thing[CONF_NAME], hass=hass))
+            elif sure_type == SureThingID.PET.name:
+                hass.data[DATA_SURE_PETCARE][
+                    sure_type][sure_id] = await surepy.get_pet_data(sure_id)
+                entities.append(Pet(sure_id, thing[CONF_NAME], hass=hass))
+
+    async_add_entities(entities, True)
+
+
+class SurePetcareBinarySensor(BinarySensorDevice):
+    """A binary sensor implementation for Sure Petcare Entities."""
+
+    def __init__(self, _id: int, name: str,
+                 icon=None, device_class=None, sure_type=None, hass=None):
+        """Initialize a Sure Petcare binary sensor."""
+        self._id = _id
+        self._hass = hass
+        self._name = name
+        self._unit_of_measurement = "%"
+        self._icon = icon
+        self._device_class = device_class
+
+        self._household_id = hass.data[DATA_SURE_PETCARE][CONF_HOUSEHOLD_ID]
+        self._sure_type = sure_type
+
+        self._state = dict()
+        self._data = hass.data[DATA_SURE_PETCARE][sure_type]
+
+    @property
+    def is_on(self):
+        """Return true if entity is on/unlocked."""
+        return bool(self._state)
+
+    @property
+    def should_poll(self):
+        """Return true."""
+        return False
+
+    @property
+    def name(self):
+        """Return the name of the device if any."""
+        return self._name
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes of the device."""
+        return self._state
+
+    @property
+    def device_class(self):
+        """Return the device class."""
+        return (DEFAULT_DEVICE_CLASS
+                if not self._device_class
+                else self._device_class)
+
+    @property
+    def icon(self):
+        """Return the device class."""
+        return DEFAULT_ICON if not self._icon else self._icon
+
+    @property
+    def unique_id(self):
+        """Return an unique ID."""
+        return "{}-{}".format(self._household_id, self._id)
+
+    async def async_update(self):
+        """Get the latest data and update the state."""
+        try:
+            self._state = self._hass.data[
+                DATA_SURE_PETCARE][self._sure_type][self._id]
+        except (AttributeError, KeyError, TypeError) as error:
```

Comment:
When would these errors happen?

Reply:
!? when one of the three listed exception occurs?!

Do you mean "these things should all be catched before"?

Reply:
We're looking up a nested dictionary that we've created ourselves and that we have control over. There should be no risk of these errors unless there's a bug.